pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/segment/SegmentSizeBasedFlushThresholdUpdater.java

```diff
@@ -0,0 +1,147 @@
+ */
+public class SegmentSizeBasedFlushThresholdUpdater implements FlushThresholdUpdater {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SegmentSizeBasedFlushThresholdUpdater.class);
+
+  private static final long IDEAL_SEGMENT_SIZE_BYTES = 200 * 1024 * 1024;
+  /** Below this size, we double the rows threshold */
+  private static final double OPTIMAL_SEGMENT_SIZE_BYTES_MIN = IDEAL_SEGMENT_SIZE_BYTES / 2;
+  /** Above this size we half the row threshold */
+  private static final double OPTIMAL_SEGMENT_SIZE_BYTES_MAX = IDEAL_SEGMENT_SIZE_BYTES * 1.5;
+  private static final int INITIAL_ROWS_THRESHOLD = 100_000;
+
+  private static final double CURRENT_SEGMENT_RATIO_WEIGHT = 0.1;
+  private static final double PREVIOUS_SEGMENT_RATIO_WEIGHT = 0.9;
+
+  @VisibleForTesting
+  long getIdealSegmentSizeBytes() {
+    return IDEAL_SEGMENT_SIZE_BYTES;
+  }
+
+  @VisibleForTesting
+  int getInitialRowsThreshold() {
+    return INITIAL_ROWS_THRESHOLD;
+  }
+
+  // num rows to segment size ratio of last committed segment for this table
+  private double _latestSegmentRowsToSizeRatio = 0;
+
+  @Override
+  public void updateFlushThreshold(@Nonnull LLCRealtimeSegmentZKMetadata newSegmentZKMetadata,
+      @Nonnull CommittingSegmentDescriptor committingSegmentDescriptor, PartitionAssignment partitionAssignment) {
+
+    LLCRealtimeSegmentZKMetadata committingSegmentZkMetadata =
+        committingSegmentDescriptor.getCommittingSegmentZkMetadata();
+    if (committingSegmentZkMetadata == null) { // first segment of the partition, hence committing segment is null
+      if (_latestSegmentRowsToSizeRatio > 0) { // new partition added case
+        LOGGER.info(
+            "Committing segment zk metadata is not available, setting rows threshold for segment {} using previous segments ratio",
+            newSegmentZKMetadata.getSegmentName());
+        long targetSegmentNumRows = (long) (IDEAL_SEGMENT_SIZE_BYTES * _latestSegmentRowsToSizeRatio);
+        targetSegmentNumRows = capNumRowsIfOverflow(targetSegmentNumRows);
+        newSegmentZKMetadata.setSizeThresholdToFlushSegment((int) targetSegmentNumRows);
+      } else {
+        LOGGER.info("Committing segment zk metadata is not available, setting default rows threshold for segment {}",
+            newSegmentZKMetadata.getSegmentName());
+        newSegmentZKMetadata.setSizeThresholdToFlushSegment(INITIAL_ROWS_THRESHOLD);
+      }
+      return;
+    }
+
+    long committingSegmentSizeBytes = committingSegmentDescriptor.getCommittingSegmentSizeBytes();
+    if (committingSegmentSizeBytes <= 0) { // repair segment case
+      LOGGER.info(
+          "Committing segment size is not available, setting thresholds for segment {} from previous segment {}",
+          newSegmentZKMetadata.getSegmentName(), committingSegmentZkMetadata.getSegmentName());
+      newSegmentZKMetadata.setSizeThresholdToFlushSegment(committingSegmentZkMetadata.getSizeThresholdToFlushSegment());
```

Comment:
good to add a comment here that we have an option to use numRowsConsumed here, but we may be bouncing back & forth between that value and twice that value (if it is much lower). 
Perhaps we can set it to numRowsConsumed * 1.1?